enhancements/openshift-console-cloud-shell.md

```diff
@@ -0,0 +1,195 @@
+apiVersion: workspaces.ecd.eclipse.org/v1alpha1
+kind: DevWorkspace
+metadata:
+  name: cloudshell
+spec:
+  components:
+    - type: dockerimage
+      id: eclipse/cloudshell/latest
+``` 
+
+As a consequence, the Cloud Shell will be dependent on the Eclipse Che DevWorkspace operator that deploys the CRD and controller to manage workspaces. But unlike Eclipse Che, in a Cloud Shell instance:
+- Eclipse Che server components are not deployed: `wsmaster`, `registries`, `keycloak` and `postgres`. Instanciation of `DevWorkspaces` is a responsibility of the OpenShift Console.
+- Eclipse Che and the Cloud Shell are both deployed and controlled using operators but those are distinct ones.
+- Theia (or other editors) based workspaces are not supported. Only Eclipse Che workspaces with editor of type CloudShell are supported.
+
+#### Authentication and Authorization
+
+The Terminal server endpoint is exposed through a TLS route. That's the endpoint that the clients running on the browser side (xterm.js) will connect to.
+
+![](https://i.imgur.com/XWcgHeI.png)
+
+The endpoint is secured through an [OpenShift OAuth Proxy](https://github.com/openshift/oauth-proxy) sidecar. The client includes the user OpenShift token in its requests HTTP header. Only authenticated OpenShift users that satisfy the RBAC policy requirements are able to send requests to the Terminal server.
+
+In order to prevent unauthorized request to the Terminal server (coming from inside the OpenShift cluster) a [Network Policy](https://kubernetes.io/docs/concepts/services-networking/network-policies/) is used.
+
+#### Cloud Shell Pod Exec Denial Policy
+
+Users tokens are injected in the Tooling container. That's required for `oc` or `odo` to work properly. As a consequence access to the Tooling container should be allowed **only to the user of the Cloud Shell instace**. That's something that the authorization mechanism based on RBAC seen above is not able to garantee: users with editor privileges at cluster scope will be able to exec into the Pod.
+
+To prevent exec from users other then the Cloud Shell owner a [ValidatingAdmissionWebhook](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/#validatingadmissionwebhook) is used.
+
+![](https://i.imgur.com/gIqTnqP.png)
+
+The Validating Webhook mechanism described above compares the user in the `AdmissionReview` request with the user that has requested the creation of the DevWorkspace CR associated to the Cloud Shell Pod. If users match the request is allowed, otherwise it's refused.
+
+This Validating Webhook mechanism requires other dynamic admission controllers to work properly:
+
+- A Mutating Admission Webhook that adds a user attribute to the DevWorkspace CR at creation time
+- A Validating Admission Webhook that block changes to the user attribute of a DevWorkspace CR
+- A Validating Admission Webhook that block changes to the Cloud Shell Pod annotation that stores the reference to the DevWorkspace CR.
```

Comment:
Could we simplify down to a least common denominator of a user reference?  A webhook on pods could actually protect an annotation (making it set-once) indicating user and then prevent other execs.  It wouldn't preclude you embedding it in another resource type, but it makes your footprint on kube smaller.

Reply:
Do you mean something like: 
- the only Workspace CR has creator annotation;
- deny modifing OwnerReference for deployment and pods;
- no need to validate deployment or pods with webhooks;
- ValidatingWebhook which check if it's try to connect to a workspace-related pod, if yes - finds to which workspace CR instance this pod belongs to, use creator from there.

If yes - we considered this option but it would make validating webhook check slower, and since we're not able to filter CreateExecOptions with object/label selectors - it would mean that every exec would be slower.

Please elaborate more if you meant something else.